### PR TITLE
[TD]Gdt reference2

### PR DESCRIPTION
**Pull request by lidiriel**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by fjullien** [comment]
Please review you commit history.
You should squash commits into logical changes.

File: src/Mod/TechDraw/App/DrawPage.cpp [hunk 1]
```diff
@@ -265,9 +266,10 @@ int DrawPage::addView(App::DocumentObject *docObj)
         return -1;
     DrawView* view = static_cast<DrawView*>(docObj);
 
-      //position all new views in center of Page (exceptDVDimension)
+      //position all new views in center of Page (except DVDimension, DVBalloon, DVGDTReference)
     if (!docObj->isDerivedFrom(TechDraw::DrawViewDimension::getClassTypeId()) &&
-        !docObj->isDerivedFrom(TechDraw::DrawViewBalloon::getClassTypeId())) {
+        !docObj->isDerivedFrom(TechDraw::DrawViewBalloon::getClassTypeId()) &&
+		!docObj->isDerivedFrom(TechDraw::DrawViewGDTReference::getClassTypeId())) {
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawPage.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/App/DrawViewGDTReference.cpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+//===========================================================================
+// DrawViewGDTReference
+//===========================================================================
+
+PROPERTY_SOURCE(TechDraw::DrawViewGDTReference, TechDraw::DrawView)
+
+const char* DrawViewGDTReference::TypeEnums[]= {"Edge",
+                                         	 "Cosmetic",
+                                             "FeatureFrame",
+											 NULL};
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawViewGDTReference.cpp]: Bad spacing, use spaces, not tabs.

File: src/Mod/TechDraw/App/DrawViewGDTReference.cpp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+DrawViewGDTReference::DrawViewGDTReference(void)
+{
+	ADD_PROPERTY_TYPE(References2D,(0,0),"",(App::Prop_None),"Projected Geometry References");
+	References2D.setScope(App::LinkScope::Global);
+
+	Type.setEnums(TypeEnums);     // object referenced type : Edge, Cosmetic line, Feature frame.
+	ADD_PROPERTY(Type,((long)0));
+	Type.setStatus(App::Property::ReadOnly,true);
+
+	ADD_PROPERTY_TYPE(Text,("A"),"",App::Prop_None,"The text to be displayed");
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawViewGDTReference.cpp]: Bad spacing. Use spaces, not tabs.

File: src/Mod/TechDraw/App/DrawViewGDTReference.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+{
+	execute();
+}
+
+short DrawViewGDTReference::mustExecute() const
+{
+    bool result = 0;
+    if (!isRestoring()) {
+        result = (	References2D.isTouched() ||
+        			Text.isTouched()	);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawViewGDTReference.cpp]: Bad spacing. Use spaces, not tabs. Why would you insert a tab before and after parentheses?

File: src/Mod/TechDraw/App/DrawViewGDTReference.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	            return App::DocumentObject::StdReturn;
+		} else {
+			Base::Console().Warning("%s has no 2D References\n", getNameInDocument());
+	    }
+		return App::DocumentObject::StdReturn;
+	}
+	//can't do anything until Source has geometry
+	if (!getViewPart()->hasGeometry()) {                              //happens when loading saved document
+		if (isRestoring() ||
+				getDocument()->testStatus(App::Document::Status::Restoring)) {
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawViewGDTReference.cpp]: Bad spacing all over here. Use spaces, not tabs.

File: src/Mod/TechDraw/App/DrawViewGDTReference.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+public:
+    /// Constructor
+    DrawViewGDTReference();
+    virtual ~DrawViewGDTReference();
+
+    App::PropertyLinkSubList	References2D;
+    App::PropertyEnumeration	Type;  // Edge, Cosmetic, Feature frame
+    App::PropertyString			Text;
+    App::PropertyFloat			SymbolScale;
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawViewGDTReference.h]: Bad spacing. Use spaces, not tabs.

File: src/Mod/TechDraw/App/DrawViewPart.cpp [hunk 7]
```diff
@@ -917,6 +933,22 @@ void DrawViewPart::unsetupObject()
...
+    page = findParentPage();
+    if (page != nullptr) {
+    	std::vector<TechDraw::DrawViewGDTReference*> refs = getGDTReferences();
+    	std::vector<TechDraw::DrawViewGDTReference*>::iterator it3 = refs.begin();
+    	for (; it3 != refs.end(); it3++) {
+    		page->removeView(*it3);
+    		const char* name = (*it3)->getNameInDocument();
+    		if (name) {
+    			Base::Interpreter().runStringArg("App.getDocument(\"%s\").removeObject(\"%s\")",
+    				docName.c_str(), name);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/App/DrawViewPart.cpp]: Bad spacing. Use spaces, not tabs.

File: src/Mod/TechDraw/Gui/Command.cpp [hunk 8]
```diff
@@ -729,6 +744,188 @@ bool CmdTechDrawProjectionGroup::isActive(void)
...
+    std::vector<std::string> subs;
+
+    std::string referenceType;
+    TechDraw::EdgeTypeUtil edgeType = _isValidSingleEdgeUtil(this);
+    if(	edgeType == TechDraw::EdgeTypeUtil::isHorizontal ||
+    	edgeType == TechDraw::EdgeTypeUtil::isVertical ||
+		edgeType == TechDraw::EdgeTypeUtil::isDiagonal){
+			referenceType = "Edge";
+			objs.push_back(objFeat);
+			subs.push_back(subNames[0]);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/Command.cpp]: Bad spacing. Use spaces, not tabs. Don't add tabs after or before parentheses.

File: src/Mod/TechDraw/Gui/Command.cpp [hunk 9]
```diff
@@ -729,6 +744,188 @@ bool CmdTechDrawProjectionGroup::isActive(void)
...
+    }else{
+    	QMessageBox::warning(Gui::getMainWindow(), QObject::tr("Incorrect Selection"),
+    			QObject::tr("Only horizontal, vertical and diagonal edge allowed."));
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/Command.cpp]: Bad spacing. Use spaces, not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 10]
```diff
@@ -0,0 +1,672 @@
...
+//TODO: hide the Qt coord system (+y down).  
+
+using namespace TechDraw;
+using namespace TechDrawGui;
+
+QGIReferenceLabel::QGIReferenceLabel(QGIViewGDTReference * parent)
+{
+	this->parent = parent;
+
+	posX = 0;
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Bad spacing. Use sapces, not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 11]
```diff
@@ -0,0 +1,672 @@
...
+void QGIReferenceLabel::setLabelCenter()
+{
+    //set label center (posX,posY) frome the given Qt position (top,left) == (x(),y())
+    posX = x() + m_labelText->boundingRect().width() / 2.;
+    posY = y() + m_labelText->boundingRect().height() / 2.;
+}
+
+void QGIReferenceLabel::setRotateAngle(double angle)
+{
+	m_angle = angle;
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 12]
```diff
@@ -0,0 +1,672 @@
...
+        Q_EMIT dragFinished();
+    }
+
+    QGraphicsItem::mouseReleaseEvent(event);
+}
+
+void QGIReferenceLabel::mouseDoubleClickEvent(QGraphicsSceneMouseEvent * event)
+{
+	// TODO nothing doing : no task managment
+	Q_UNUSED(event);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+{
+    m_colNormal = c;
+    m_labelText->setColor(m_colNormal);
+}
+
+void QGIReferenceLabel::rotate(void)
+{
+	QPointF centerPt = boundingRect().center();
+	setTransformOriginPoint(centerPt);
+	setRotation(m_angle);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 14]
```diff
@@ -0,0 +1,672 @@
...
+    		referenceLabel, SIGNAL(dragFinished()),
+        this  , SLOT  (referenceLabelDragFinished()));
+
+    QObject::connect(
+    		referenceLabel, SIGNAL(selected(bool)),
+        this  , SLOT  (select(bool)));
+
+    QObject::connect(
+    		referenceLabel, SIGNAL(hover(bool)),
+        this  , SLOT  (hover(bool)));
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Bad spacing. Use spaces, not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+
+    setZValue(ZVALUE::DIMENSION);                    //note: this won't paint dimensions over another View if it stacks
+                                                     //above this Dimension's parent view.   need Layers?
+
+}
+
+Base::Vector3d QGIViewGDTReference::calculateCenter(TechDraw::PointPair & segment){
+	Base::Vector3d first = segment.first;
+	Base::Vector3d second = segment.second;
+	return (first+second)/2.0;
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces, not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 16]
```diff
@@ -0,0 +1,672 @@
...
+
+	m_arrowAngle = atan2(-perp.y, -perp.x) * 180 / M_PI;
+	m_linkDir = perp;
+
+	float x = Rez::guiX(labelCenter.x);
+	float y = Rez::guiX(labelCenter.y);
+	referenceLabel->setPosFromCenter(x, y);
+	Base::Vector2d result = fromQtApp(origin);
+
+	return result;
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 17]
```diff
@@ -0,0 +1,672 @@
...
+    auto viewProvider = static_cast<ViewProviderGDTReference*>(getViewProvider(getViewObject()));
+    if ( viewProvider == nullptr ) {
+        return;
+    }
+
+    if(	update ||
+    	reference->X.isTouched() || reference->Y.isTouched() ||
+		reference->Text.isTouched() ||
+		viewProvider->Font.isTouched() ||
+		viewProvider->LineWidth.isTouched()){
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs. Don't use tabs after or before parentheses.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 18]
```diff
@@ -0,0 +1,672 @@
...
+	Base::Vector2d startPt = fromQtApp(edge.first);
+	intersectionPt = computePerpendicularIntersection(startPt, labelCenter, m_lineAngle);
+
+	Base::Vector2d perp = (labelCenter-intersectionPt);
+	perp.Normalize();
+	m_linkDir = Base::Vector3d(perp.x, -perp.y);
+	m_arrowAngle = atan2(perp.y, -perp.x) * 180 / M_PI;
+
+	reference->X.setValue(intersectionPt.x);
+	reference->Y.setValue(intersectionPt.y);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 19]
```diff
@@ -0,0 +1,672 @@
...
+}
+
+void QGIViewGDTReference::draw()
+{
+    draw_modifier(false);
+}
+
+void QGIViewGDTReference::draw_modifier(bool modifier)
+{
+	Q_UNUSED(modifier);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 20]
```diff
@@ -0,0 +1,672 @@
...
+        parentItem()->update();
+    } else {
+        Base::Console().Log("INFO - QGIViewGDTReference::draw_modifier - no parent to update\n");
+    }
+
+}
+
+void QGIViewGDTReference::drawBorder(void)
+{
+	// Nothing doing !
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp [hunk 21]
```diff
@@ -0,0 +1,672 @@
...
+    double svgLineFactor = 3.0;                     //magic number.  should be a setting somewhere.
+    referenceLines->setWidth(m_lineWidth/svgLineFactor);
+    referenceShape->setWidth(m_lineWidth/svgLineFactor);
+    referenceArrow->setWidth(referenceArrow->getWidth()/svgLineFactor);
+}
+
+void QGIViewGDTReference::setPens(void)
+{
+	referenceLines->setWidth(m_lineWidth);
+	referenceShape->setWidth(m_lineWidth);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.h [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+class QGIArrow;
+class QGIDimLines;
+class QGIViewGDTReference;
+
+class QGIReferenceLabel : public QGraphicsObject
+{
+Q_OBJECT
+
+public:
+	QGIReferenceLabel(QGIViewGDTReference* parent);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.h]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGIViewGDTReference.h [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    Base::Vector3d calculateCenter(TechDraw::PointPair & segment);
+    Base::Vector2d labelPlacementAndRotation(TechDraw::PointPair & segment, double distance);
+    // TODO move to utils (using in QGIViewDimension)
+    double getIsoStandardLinePlacement(double labelAngle);
+    static inline double toDeg(double a) { return a*180/M_PI; }
+    static inline Base::Vector2d fromQtApp(const Base::Vector3d &v) { return Base::Vector2d(v.x, -v.y); }
+    static inline double toQtDeg(double a) { return -a*180.0/M_PI; }
+    static inline double toQtRad(double a) { return -a*M_PI/180.0; }
+    Base::Vector2d computePerpendicularIntersection(const Base::Vector2d &linePoint,
+    		const Base::Vector2d &perpendicularPoint, double lineAngle);
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGIViewGDTReference.h]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGVPage.cpp [hunk 24]
```diff
@@ -622,6 +654,7 @@ QGIView * QGVPage::findParent(QGIView *view) const
         std::vector<App::DocumentObject *> objs = dim->References2D.getValues();
 
         if(objs.size() > 0) {
+        	// TODO correct this line ?
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGVPage.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/QGVPage.cpp [hunk 25]
```diff
@@ -632,6 +665,22 @@ QGIView * QGVPage::findParent(QGIView *view) const
...
 
+    //If type is gdtreference we check references2D first
+    TechDraw::DrawViewGDTReference *ref = 0;
+    ref = dynamic_cast<TechDraw::DrawViewGDTReference *>(myFeat);
+
+    if(ref) {
+    	std::vector<App::DocumentObject *> objs = ref->References2D.getValues();
+        if(objs.size() > 0) {
+        	// Attach the GDTReference to the first object's group
+        	for(std::vector<QGIView *>::const_iterator it = qviews.begin(); it != qviews.end(); ++it) {
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/QGVPage.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/ViewProviderPage.cpp [hunk 26]
```diff
@@ -282,6 +283,7 @@ std::vector<App::DocumentObject*> ViewProviderPage::claimChildren(void) const
     // for Page, valid children are any View except: DrawProjGroupItem
     //                                               DrawViewDimension
     //                                               DrawViewBalloon
+    //												 DrawViewGDTReference
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/ViewProviderPage.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/ViewProviderPage.cpp [hunk 27]
```diff
@@ -313,6 +315,7 @@ std::vector<App::DocumentObject*> ViewProviderPage::claimChildren(void) const
              docObj->isDerivedFrom(TechDraw::DrawRichAnno::getClassTypeId())         ||
              docObj->isDerivedFrom(TechDraw::DrawLeaderLine::getClassTypeId())       ||
              docObj->isDerivedFrom(TechDraw::DrawWeldSymbol::getClassTypeId())       ||
+			 docObj->isDerivedFrom(TechDraw::DrawViewGDTReference::getClassTypeId()) ||
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/ViewProviderPage.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/ViewProviderViewPart.cpp [hunk 28]
```diff
@@ -169,6 +170,7 @@ std::vector<App::DocumentObject*> ViewProviderViewPart::claimChildren(void) cons
     //    - Hatches
     //    - GeomHatches
     //    - Leaders
+	//    - GDTReference
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/ViewProviderViewPart.cpp]: Bad spacing. Use spaces not tabs.

File: src/Mod/TechDraw/Gui/ViewProviderViewPart.cpp [hunk 29]
```diff
@@ -197,6 +199,8 @@ std::vector<App::DocumentObject*> ViewProviderViewPart::claimChildren(void) cons
               temp.push_back((*it));
           } else if ((*it)->getTypeId().isDerivedFrom(TechDraw::DrawLeaderLine::getClassTypeId())) {
               temp.push_back((*it));
+      	  } else if ((*it)->getTypeId().isDerivedFrom(TechDraw::DrawViewGDTReference::getClassTypeId())) {
+      		  temp.push_back((*it));
```
Comment (vocx-fc) [inline src/Mod/TechDraw/Gui/ViewProviderViewPart.cpp]: Bad spacing. Use spaces not tabs.

**Comment by vocx-fc** [comment]
I didn't check the code for correctness, but the spacing is pretty weird. It's inconsistent, mixing spaces and tabs, even within the same function.

And I agree with fjullien, there seem to be too many commits. I don't think you have to squash all into only one but maybe you could squash some to reduce the number of commits to some logical sections.

**Comment by WandererFan** [comment]
vocx: 31 comments!  could have been just 1 for fix spacing and 1 for squashing. :)

**Comment by WandererFan** [comment]
lidiriel: this commit scheme is good for development, not so good for review and repository update.  Any chance you can squash related commits so it isn't so hard to see what you've changed?
Thanks,
wf

**Comment by vocx-fc** [comment]
This way he'll know exactly which lines I'm talking about, plus the ones that I didn't catch.

**Comment by lidiriel** [comment]
Thanks for yours review. I will rewrite code to correct : space in indent, squash many commit into less than 5 or 6. I think it is possible.
I repost a message when i will ready for a new pull request.

byr

**Comment by vocx-fc** [comment]
Looks fine. Now I suggest squashing the commits and force pushing.

**Comment by WandererFan** [comment]
It would make sense to cancel this PR and make a new on when you are ready to merge.
WF

**Comment by lidiriel** [comment]
ok. i cancel it for this time. I have corrected several elements. But i must implement  auto increment of reference value.

**Pull request closed without merging**